File: python/net.py
# %%
import torch
from torch import nn
# %%
class Lattice():
    def __init__(self, L, dim=2):
        self.L = L
        self.d = dim
        
    def init_phi(self):
        return torch.zeros((self.L,)*self.d)
    
    def dist_fun(self):
        dist = torch.zeros((self.d,) + (self.L,)*self.d)
        for k in range(self.d):
            for i in range(self.L):
                dist[k] += i / self.L
        return dist
    
class Net(nn.Module):
    def __init__(self, d, M, depth=3, width=16, agg="sum"):
        super().__init__()
        assert depth > 1
        
        layers1 = [nn.Linear(d, width),
                   nn.Tanh()]
        for _ in range(depth-1):
            layers1.append(nn.Linear(width, width))
            layers1.append(nn.ReLU())

        self.net1 = nn.Sequential(*layers1)
        
        self.aggregate = torch.sum
        
        layers2 = []
        for _ in range(depth-1):
            layers2.append(nn.Linear(width, width))
            layers2.append(nn.ReLU())
        layers2.append(nn.Linear(width, 3*M))
        
        self.net2 = nn.Sequential(*layers2)
        
    def forward(self, x):
        x2 = self.net1(x)
        x2 = self.aggregate(x2, dim=1)
        params = self.net2(x2)
        w, mu, logsig = torch.split(params, 3, dim=2)
        sig = torch.exp(0.5 * logsig)
        w = torch.softmax(w, dim=2)
        # Standard Gaussian random numbers
        rng = torch.randn_like(mu)
        
        return torch.sum(w*(mu + sig*rng), dim=2)
        
def sample(net, L, batch):
    phi_pad = torch.zeros((batch,) + (L+1,)*2).detach()
    
    for i in range(1, L+1):
        for j in range(1, L+1):
            phi_neigh = torch.stack(
                [phi_pad[:, i-1, j], phi_pad[:, i, j-1]], 1)
            # (batch, 2) array
            dist = torch.stack(
                [torch.full((batch,), (i-1)/L),
                 torch.full((batch,), (j-1)/L)], 1)
            # (batch, 2) array
            flag = torch.stack(
                [torch.full((batch,), 1 if i==1 else 0),
                 torch.full((batch,), 1 if j==1 else 0)], 1)
            # (batch, 2) array
            
            inp = torch.stack([phi_neigh, dist, flag], 2)
            phi_pad[:, i, j] = net(inp)
            

File: julia/net.jl
using Flux: randn32
using Flux, ArgCheck

struct Lattice
    L::Int
    a::AbstractFloat
    dim::Int
end

# Fix dim=2 (1 space + 1 time) for now
Lattice(L, a) = Lattice(L, a, 2)

struct DeepSets
    fc1::Chain
    fc2::Chain
    agg
end

function DeepSets(depth=3, width=32)
    @argcheck depth > 1
    layers1 = [Dense(2 => width, tanh)]
    append!(layers1, [Dense(width => width, relu) 
                      for i=1:depth-1])
    fc1 = Chain(layers1...) # NN of given depth and width
    
    layers2 = [Dense(width => width, relu)
               for i=1:depth-1]
    append!(layers2, [Dense(width => 2)])
    fc2 = Chain(layers2...) # NN of given depth and width
    
    agg = sum # aggregating function for DeepSets
    return DeepSets(fc1, fc2, agg)
end

"""
Applies deepsets network and outputs parameters of 
mixed Gaussian distribution
"""
function call(net::DeepSets, x) # size(x) = (3, 2, batch)
    x2 = net.fc1(x)
    # Aggregate along set dimension 2
    x2 = agg(x2; dims=2)[:, 1, :]
    # Obtain distribution params after acting 2nd net
    dist_param = net.fc2(x2)
    # Split between mean and variance
    μ, σ = Flux.chunk(dist_param, 2; dims=1)
    σ = exp.(logσ) # Make positive
    ϵ = randn32(size(μ)...)
    sample = (μ .+ ϵ.*σ)[1, :] # Gaussian sample, size: (batch,)
    logprob = (-log.(σ) - (ϵ^2)./2)[1, :]
end

function sample(lattice::Lattice, net::DeepSets, batch)
    L = lattice.L
    T = Float32
    # Lattice with 1 zero padding on every dim
    ϕ_pad = zeros(L+1, L+1, batch)
    # T store log probabilities of every sampled value
    ϕ_lp = zeros(L, L, batch)
    for t in 2:L+1, x in 2:L+1
        # Neighbour ϕ values at every lattice position
        ϕ_neigh = Flux.stack(
            [ϕ_pad[x-1, t, :], ϕ_pad[x, t-1, :]]; 
        dims=1)
        # Distance from zero for every dimension
        dist = Flux.stack(
            [fill(T((x-1)/L), batch), 
             fill(T((t-1)/L), batch)]; dims=1)
        # Flags if neighbour padded
        flag = Flux.stack(
            [fill(x==2 ? 1 : 0, batch),
             fill(t==2 ? 1 : 0, batch)]; dims=1)
        # Stacking them for NN input
        inp = Flux.stack([ϕ_neigh, dist, flag]; dims=1)
        # Get sample and logprob from NN
        sample, logprob = call(net, inp) #(M, batch) arrays
        ϕ_pad[x, t, :] = sample
        ϕ_lp[x-1, t-1, :] = logprob
    end
end